src/main/java/cam72cam/immersiverailroading/util/CAML.java

```diff
@@ -0,0 +1,172 @@
+ *          string = Any sequence of characters that is optionally wrapped in "quotation marks"
+ *      types =
+ *          # There are two types in CAML
+ *          blocks =
+ *              description = A block is a named container
+ *              format = A block is started with a key at the current indentation level
+ *              contents = Any subsequent lines at a deeper indentation level are read and then added to the block as their respective types.
+ *
+ *          primitive = A key followed by a value
+ *
+ * */
+public class CAML {
+    private static final Pattern base = Pattern.compile("(\\s*)(\\S+)\\s*([=:])\\s?(.*)");
+    public static DataBlock parse(InputStream stream) throws IOException {
+        List<String> lines = IOUtils.readLines(stream, StandardCharsets.UTF_8).stream()
+                .map(s -> s.replaceFirst("#.*", ""))
+                .filter(s -> !StringUtils.isWhitespace(s))
+                .collect(Collectors.toList());
+
+        return new Block(lines);
+    }
+
+    private static class Block implements DataBlock {
+        private final Map<String, Value> primitives = new LinkedHashMap<>();
+        private final Map<String, List<Value>> primitiveSets = new LinkedHashMap<>();
+        private final Map<String, DataBlock> blocks = new LinkedHashMap<>();
+        private final Map<String, List<DataBlock>> blockSets = new LinkedHashMap<>();
+
+        private Block(List<String> lines) throws IOException {
+            String spaces = null;
+            while (!lines.isEmpty()) {
+                String line = lines.get(0);
+                Matcher m = base.matcher(line);
+                if (!m.matches()) {
+                    throw new IOException(String.format("Invalid Block line '%s'", line));
+                }
+
+                String pre = m.group(1);
+                String key = m.group(2);
+                String mod = m.group(3);
+                String val = m.group(4);
+
+                if (spaces == null) {
+                    spaces = pre;
+                }
+
+                if (!pre.startsWith(spaces) && !spaces.startsWith(pre)) {
+                    throw new IOException(String.format("Invalid Block line '%s' mismatched indentation '%s' vs '%s'", line, spaces, pre));
+                }
+
+                if (spaces.length() > pre.length()) {
+                    // Reduced indentation
+                    return;
+                }
+                if (!spaces.equals(pre)) {
+                    throw new IOException(String.format("Invalid Block line '%s' invalid indentation '%s' vs '%s'", line, spaces, pre));
+                }
+
+                lines.remove(0);
+
+                String trimmed = val.trim();
+                if (trimmed.startsWith("\"") && trimmed.endsWith("\"")) {
+                    val = val.substring(1, val.length()-1);
+                }
+
+                if (StringUtils.isWhitespace(val)) {
+                    Block block = new Block(lines);
```

Comment:
yikes, a recursive ctor?

Reply:
Yep, I could break out parsing / constructing.  This was the simplest path at the time.  It's not terrible IMO.